File: trunk/tests/test_with_Mathematica.c
/*****************************************************************************
 *                                                                           *
 *          UNURAN -- Universal Non-Uniform Random number generator          *
 *                                                                           *
 *****************************************************************************
     $Id$ 
 *****************************************************************************
 *                                                                           *
 *  Compare CDF, PDF and derivatives of PDF of varios distributions          *
 *  with Mathematica output.                                                 *
 *                                                                           *
 *****************************************************************************/

/*---------------------------------------------------------------------------*/

/*  #define DEBUG 1 */

/*---------------------------------------------------------------------------*/

#include "test_with_Mathematica.h"

/*---------------------------------------------------------------------------*/

int test_cont_cdf_pdf( UNUR_DISTR *distr, char *datafile, double max_diff )
     /*----------------------------------------------------------------------*/
     /* test CDF, PDF and derivative of PDF by comparing to data             */
     /* read from file created by Mathematica.                               */
     /*                                                                      */
     /* the ordering in this file is:                                        */
     /*   n   par[1] par[2] ... par[n]   x   CDF   PDF   dPDF                */
     /*                                                                      */
     /* parameters:                                                          */
     /*   distr    ... pointer to distribution object                        */
     /*   datafile ... name of file where data are stored                    */
     /*   max_diff ... maximal allowed difference                            */
     /*                                                                      */
     /* return:                                                              */
     /*   1 ... if test was successful                                       */
     /*   0 ... difference too large                                         */
     /*----------------------------------------------------------------------*/
{
#define BUFSIZE      1024       /* size of line buffer */
#define MAX_FPARAMS  10         /* maximal number of parameters for distribution */

  FILE *fp;                     /* file handle for input file */
  int lineno;                   /* line number in file */
  char buffer[BUFSIZE];         /* buffer for reading line */
  char *ptr_buffer;             /* pointer into buffer */

  int n_fparams;                /* number of parameters for distribution */
  double fparams[10];           /* array for parameters for distribution */
  double x;                     /* argument */
  double CDF_e, PDF_e, dPDF_e;  /* expected values for CDF, PDF and derivative of PDF at x */
  double CDF_o, PDF_o, dPDF_o;  /* observed values for CDF, PDF and derivative of PDF at x */
  double CDF_d, PDF_d, dPDF_d;  /* differences between observed and expected values */

  double CDF_md, PDF_md, dPDF_md;  /* maximal difference (absolute) */

  int i;

  /* initialize */
  CDF_md = PDF_md = dPDF_md = 0.;

  /* print info on screen */
  printf("%s ...\n",datafile);

  /* open file for reading */
  if ( (fp = fopen(datafile,"r")) == NULL ) {
    printf("ERROR: could not open file %s \n",datafile);
    return(-1.);
  }

  /* read file */
  for (lineno = 0;; ++lineno) {

    /* read next line */
    fgets(buffer, BUFSIZE-1, fp);
    if (feof(fp)) break;
    ptr_buffer = buffer;

    /* get number of parameters */
    n_fparams = strtol(buffer, &ptr_buffer, 10);
    if (n_fparams < 0 || n_fparams >= MAX_FPARAMS) {
      printf("ERROR: invalid number of parameters for distribution: %d \n",n_fparams);
      return(-1.);
    }

    /* read parameters */
    for (i=0; i<n_fparams; i++) {
      fparams[i] = strtod( ptr_buffer, &ptr_buffer );
    }

    /* read argument */
    x = strtod( ptr_buffer, &ptr_buffer );

    /* read CDF */
    CDF_e = strtod( ptr_buffer, &ptr_buffer );

    /* read PDF */
    PDF_e = strtod( ptr_buffer, &ptr_buffer );

    /* read dPDF */
    dPDF_e = strtod( ptr_buffer, &ptr_buffer );

#ifdef DEBUG
    /* print data */
    printf("n_fparams = %d\n", n_fparams);

    /* print parameters */
    for (i=0; i<n_fparams; i++)
      printf("\t%d: %g\n",i,fparams[i]);

    /* print argument x */
    printf("expected x = %g:\t",x);

    /* print CDF, PDF and derivative of PDF at x */
    printf("%g, %g, %g\n",CDF_e,PDF_e,dPDF_e);

#endif

    /* set parameters for distribution */
    unur_distr_cont_set_pdfparams(distr,fparams,n_fparams);
    unur_distr_cont_upd_pdfarea(distr); 

    /* compute CDF, PDF and derivative of PDF at x */
    CDF_o = unur_distr_cont_eval_cdf(x, distr);
    PDF_o = unur_distr_cont_eval_pdf(x, distr);
    dPDF_o = unur_distr_cont_eval_dpdf(x, distr);

    /* compute differnces */
    CDF_d = CDF_o - CDF_e;
    PDF_d = PDF_o - PDF_e;
    dPDF_d = dPDF_o - dPDF_e;

#ifdef DEBUG
    /* print argument x */
    printf("observed x = %g:\t",x);

    /* print CDF, PDF and derivative of PDF at x */
    printf("%g, %g, %g\n",CDF_o,PDF_o,dPDF_o);

    /* print differences */
    printf("diff     x = %g:\t",x);
    printf("%g, %g, %g\n\n",CDF_d, PDF_d, dPDF_d);
#endif
    
    /* compare differences */
    CDF_d = fabs(CDF_d);
    PDF_d = fabs(PDF_d);
    dPDF_d = fabs(dPDF_d);
    dPDF_e = fabs(dPDF_e);

    if (CDF_d > CDF_md)   CDF_md = CDF_d;
    if (PDF_d > PDF_md)   PDF_md = PDF_d;
    if (dPDF_d > dPDF_md) dPDF_md = dPDF_d;

  }

  /* close file handle */
  fclose(fp);

  /* print info on screen */
  printf("\tmaximal difference:\n");
  printf("\t\tCDF  = %g\n",CDF_md);
  printf("\t\tPDF  = %g\n",PDF_md);
  printf("\t\tdPDF = %g\n",dPDF_md);

  return 0;

#undef BUFSIZE
#undef MAX_FPARAMS
} 

/*---------------------------------------------------------------------------*/

File: trunk/tests/test_with_Mathematica.m
(*****************************************************************************
 *                                                                           *
 *          UNURAN -- Universal Non-Uniform Random number generator          *
 *                                                                           *
 *****************************************************************************
     $Id$ 
 *****************************************************************************
 *                                                                           *
 *  Create file with results of CDF, PDF and derivatives of PDF              *
 *  for various distributions and make C test file.                          *
 *                                                                           *
 *****************************************************************************)

(* === Set Constants ========================================================*)

(* name of C file for running tests *)
RunFileName = "run_test_with_Mathematica.c";

(* sample size for tests *)
RunSampleSize = 5;   

(* === Load statistics packages =============================================*)
<<Statistics`ContinuousDistributions`
<<Statistics`DiscreteDistributions`

(* === Declare additional distributions ==================================== *)

(* --- Defaults (to avoid error with C program) ---------------------------- *)

CDF[___] := 0.;
PDF[___] := 0.;

(* --- Powerexponential Distribution --------------------------------------- *)

PowerexponentialDistribution/: Domain[PowerexponentialDistribution[___]] := Interval[{-Infinity, Infinity}]

PowerexponentialDistribution/: PDF[PowerexponentialDistribution[r_], x_] :=
	If [ x>0, 
		Exp[-x^r]/(2 Gamma[1+1/r]),
	(* Else *)
		Exp[-(-x)^r]/(2 Gamma[1+1/r]) ];

PowerexponentialDistribution/: CDF[PowerexponentialDistribution[r_], x_] := 
	If [ x<0,
		Integrate[ Exp[-(-t)^r], {t,-Infinity,x}] / (2 Gamma[1+1/r]),
	(* Else *)
		1/2 + Integrate[Exp[-(t)^r], {t,0,x}] / (2 Gamma[1+1/r]) ];

PowerexponentialDistribution/: Random[PowerexponentialDistribution[r_]] := 
	(* Warning!! This not a generator for the Powerexponential Distribution!!       *)
	(* FindRoot[ PDF[PowerexponentialDistribution[r],x] == Random[],{x,0.3}][[1,2]] *)
	(*  is correct but very slow. So we use a Laplace random variate instead !!     *)
	If[ Random[] < 0.5, 1, -1] * If[ Random[] < Max[0.3,(1-1/r)], Random[], 1-Log[Random[]]/r ];

(* --- Laplace Distribution ------------------------------------------------ *)

(* Mathematica's implementation of the Laplace Distribution is wrong!!       *)
(* We have to fix it!                                                        *)

Unprotect[LaplaceDistribution];
Clear[LaplaceDistribution];

LaplaceDistribution/: Domain[LaplaceDistribution[___]] := Interval[{-Infinity, Infinity}]

LaplaceDistribution/: PDF[LaplaceDistribution[m_,s_], x_] :=
	If [ x-m < 0,
		Exp[(x-m)/s] / (2 s),
	(* Else *)
		Exp[(-x+m)/s] / (2 s) ];

LaplaceDistribution/: CDF[LaplaceDistribution[m_,s_], x_] :=
	If [ x-m < 0,
		Integrate[ Exp[(t-m)/s] / (2 s), {t,-Infinity,x}],
	(* Else *)
		1/2 + Integrate[ Exp[(-t+m)/s] / (2 s), {t,m,x}] ];

LaplaceDistribution/: Random[LaplaceDistribution[m_,s_]] := 
	m + If[ Random[]>0.5, 1,-1 ] * s * Log[Random[]];

(* === End of Distributions ================================================ *)

(* === Format and write output ============================================= *)

(* --- Write result for parameter/argument combination into file ----------- *)

UnurTestDistrResultLine[stream_, distr_, fparams__, x_] := Module [
	(*	stream  ... output stream                                    *)
	(*   	distr   ... distribution                                     *)
	(*	fparams ... parameters for distribution                      *)
	(*	x       ... argument where `distr' should be evaluated       *)

	(* local variables *)
	{i},

	(* number of parameters *)
	WriteString[stream, Length[fparams]," "];

	(* list of parameters *)
	Table[
		WriteString[stream, CForm[ N[ fparams[[i]], $MachinePrecision+1]]," "],
		{i,1,Length[fparams]} ];

	(* argument *)
	WriteString[stream, CForm[ N[x+0.,$MachinePrecision+1] ]," "];

	(* CDF *)
	WriteString[stream, CForm[ N[ CDF[ Apply[distr,fparams], x+0. ], $MachinePrecision+1]]," "];
	
	(* PDF *)
	WriteString[stream, CForm[ N[ PDF[ Apply[distr,fparams], x+0. ], $MachinePrecision+1]]," "];
	
	(* derivative of PDF *)
	WriteString[stream, CForm[ N[ 0. + D[ PDF[ Apply[distr,fparams], t ], t] /. t->x, $MachinePrecision+1]]," "];

	(* end *)
	WriteString[stream, "\n"];
]; (* end of UnurTestDistrResultLine[] *)

(* --- make file with results for given distribution ----------------------- *)

UnurTestDistrResultFile[dname_, distr_, runfile_, fparbd__, size_] := Module [ 
	(*	dname   ... UNURAN name of distribution                      *)
	(*   	distr   ... distribution                                     *)
	(*	runfile ... output stream for C file running tests           *)
	(*	fparbd  ... bounds for parameters for distribution           *)
	(*	size    ... size of sample                                   *)
	
	(* local variables *)
	{stream, datafilename, i, j, nfparams, fparams, x},

	(* compose a file name for output *)
	datafilename = "t_distr_" <> dname <> ".data";
	Print[ datafilename ];
	
	(* open output stream *)
	stream = OpenWrite[datafilename];

	(* number of parameters for distribution *)
	nfparams = Length[fparbd];

	Do[
		(* make parameters for distribution at random *)
		fparams = 
			Table[ fparbd[[j,1]] + Round[(fparbd[[j,2]]-fparbd[[j,1]])*1000*Random[]]/1000,
			{j,nfparams} ];	

		(* make an argument at random (use importance sampling) *)
		x = Round[1000 * Random[ Apply[distr,fparams] ]] / 1000;

		(* print line with this parameters *)
		UnurTestDistrResultLine[stream,distr,fparams,x],

	{i,size} ];

	(* close output stream *)
	Close[stream];

	(* print test line into C file running tests *)

	(* list of parameters. we use the mean of upper and lower bound *)
	For [ i=0, i<nfparams, i++,
		x = 0.5 * (fparbd[[i+1,1]]+fparbd[[i+1,2]]);
		WriteString[runfile,"   fparams[",i,"] = ",x,";\n"];
	];
	WriteString[runfile,"   distr = unur_distr_",dname,"(fparams,",nfparams,");\n"];
	WriteString[runfile,"   test_cdf_pdf( distr,\"",datafilename,"\", 1. );\n"];
        WriteString[runfile,"   unur_distr_free(distr);\n\n"];

]; (* end of UnurTestDistrResultFile[] *)

(* --- Open C file for output----------------------------------------------- *)

UnurTestRunOpen[filename_] := Module [
	(*	filename ... name of output file                             *)

	(* local variables *)
	{stream},

	(* open file for writing *)
	stream = OpenWrite[filename];

	(* write file preamble *)
	WriteString[stream,"\
/***************************************************************************** \n\
 *                                                                           * \n\
 *          UNURAN -- Universal Non-Uniform Random number generator          * \n\
 *                                                                           * \n\
 *****************************************************************************/\n\
                                                                               \n\
/*---------------------------------------------------------------------------*/\n\
/*  #define DEBUG 1 */                                                         \n\
/*---------------------------------------------------------------------------*/\n\
#include \"test_with_Mathematica.h\"                                           \n\
/*---------------------------------------------------------------------------*/\n\
                                                                               \n\
int main()                                                                     \n\
{                                                                              \n\
   double fparams[UNUR_DISTR_MAXPARAMS];  /* parameters of distribution      */\n\
   UNUR_DISTR *distr;                     /* distribution object             */\n\
   FILE *UNURANLOG;                       /* unuran log file                 */\n\
                                                                               \n\
   /* open log file for unuran and set output stream for unuran messages */    \n\
   UNURANLOG = fopen( \"t_Mathematica_unuran.log\",\"w\" );                    \n\
   if (UNURANLOG == NULL) exit (-1);                                           \n\
   unur_set_stream( UNURANLOG );                                               \n\
                                                                               \n"
	];

	(* return stream handler *)
	Return[ stream ];

]; (* end of UnurTestRunOpen[] *)

(* --- Close output file -------------------------------------------------- *)

UnurTestRunClose[stream_] := Module [
	(*	stream ... output stream                                    *)

	(* local variables: none *)
	{},

	(* write file postamble *)
	WriteString[stream,"\
                                                                               \n\
   /* close log file */                                                        \n\
   fclose(UNURANLOG);                                                          \n\
                                                                               \n\
   exit (0);                                                                   \n\
}                                                                              \n"
	];

	(* close stream *)
	Close[stream];

]; (* end of UnurTestRunClose[] *)

(* === Start of Main ======================================================= *)

(* Open C file *)
runfile = UnurTestRunOpen[ RunFileName ];

(* --- List of Continuous Distributions ------------------------------------ *)

(* Beta *)
fparams = {{1.,10.},{1.,100.}};
UnurTestDistrResultFile["beta", BetaDistribution, runfile, fparams, RunSampleSize];

(* Gamma *)
fparams = {{0.5,10.},{0.01,100.}};
UnurTestDistrResultFile["gamma", GammaDistribution, runfile, fparams, RunSampleSize];

(* Laplace *)
fparams = {{-100.,100.},{0.01,100.}};
UnurTestDistrResultFile["laplace", LaplaceDistribution, runfile, fparams, RunSampleSize];

(* Normal *)
fparams = {{0.5,10.},{0.01,100.}};
UnurTestDistrResultFile["normal", NormalDistribution, runfile, fparams, RunSampleSize];

(* Powerexponential *)
fparams = {{1/1000,3}};
UnurTestDistrResultFile["powerexponential", PowerexponentialDistribution, runfile, fparams, RunSampleSize];

(* --- Done ---------------------------------------------------------------- *)

(* close C file *)
UnurTestRunClose[runfile];

(* === Exit ================================================================ *)

Quit[]







